webrtc/README.md

```diff
@@ -0,0 +1,445 @@
+   across connections by on-path observers given that WebRTC uses TLS 1.2.
+
+4. _A_ constructs _B_'s SDP offer locally based on _B_'s multiaddr. _A_
+   generates a random string and sets that string as the username (_ufrag_ or
+   _username fragment_) and password on the SDP of the remote offer. Finally _A_
+   sets the remote offer via
+   [`RTCPeerConnection.setRemoteDescription()`](https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/setRemoteDescription).
+
+5. _A_ creates a local offer via
+   [`RTCPeerConnection.createOffer()`](https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/createOffer).
+   _A_ sets the same username as password on the local offer as done in (4) on
+   the remote offer. Finally _A_ sets the modified offer via
+   [`RTCPeerConnection.setLocalDescription()`](https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/setLocalDescription).
+
+   Note that this process, oftentimes referred to as "SDP munging" is disallowed
+   by the specification, but not enforced across the major browsers (Safari,
+   Firefox, Chrome) due to use-cases in the wild. See also
+   https://bugs.chromium.org/p/chromium/issues/detail?id=823036
+
+6. _A_ establishes the connection to _B_. The random string used as a _username_
+   and _password_ in combination with the IP and port of _A_ can be used by _B_
+   to identify the connection, i.e. demultiplex incoming UDP datagrams per
+   incoming connection. _B_ uses the same random string for the username and
+   password in the STUN message from _B_ to _A_.
+
+   Note that this step requires _B_ to allocate memory for each incoming STUN
+   message from _A_. This could be leveraged for a DOS attack where _A_ is
+   sending many STUN messages with different ufrags using different UDP source
+   ports, forcing _B_ to allocate a new peer connection for each. _B_ SHOULD
+   have a rate limiting mechanism in place as a defense measure. See also
+   https://datatracker.ietf.org/doc/html/rfc5389#section-16.1.2.
+
+7. _A_ and _B_ execute the DTLS handshake as part of the standard WebRTC
+   connection establishment.
+
+   At this point _B_ does not know the TLS certificate fingerprint of _A_. Thus
+   _B_ can not verify _A_'s TLS certificate fingerprint during the DTLS
+   handshake. Instead _B_ needs to _disable certificate fingerprint
+   verification_ (see e.g. [Pion's `disableCertificateFingerprintVerification`
+   option](https://github.com/pion/webrtc/blob/360b0f1745c7244850ed638f423cda716a81cedf/settingengine.go#L62)).
+
+   On success of the DTLS handshake the connection provides confidentiality and
+   integrity but not authenticity. The latter is guaranteed through the
+   succeeding Noise handshake. See [Connection Security
+   section](#connection-security).
+
+8. Messages on the established `RTCDataChannel` are framed using the message
+   framing mechanism described in [Multiplexing](#multiplexing).
```

Comment:
What's the reason behind this choice? None of the 3 flags makes sense in the context of the handshake, as you only have a single substream, and you can just call `RTCPeerConnection.close()` if you want to abort the handshake.

Reply:
The reasons are:

- Consistency, i.e. there is only one way how we write to a data channel, namely framed with our Protobuf.
- Potential upgrade mechanism where we can embed additional fields in the Protobuf, even during the handshake.

Reply:
Resolving here. @tomaka let me know in case the above does not answer your question.